File: 449_Serialize_and_Deserialize_BST.py
# Serialization is the process of converting a data structure or object into a sequence of bits so that it can be stored in a file or memory buffer, or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

# Design an algorithm to serialize and deserialize a binary search tree. There is no restriction on how your serialization/deserialization algorithm should work. You just need to ensure that a binary search tree can be serialized to a string and this string can be deserialized to the original tree structure.

# The encoded string should be as compact as possible.

# Note: Do not use class member/global/static variables to store states. Your serialize and deserialize algorithms should be 
class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        if root.left:
            if root.right:
                return "{}({})({})".format(root.val,self.serialize(root.left),self.serialize(root.right))
            else:
                return "{}({})".format(root.val,self.serialize(root.left))
        else:
            if root.right:
                return "{}()({})".format(root.val,self.serialize(root.right))
            else:
                return "{}".format(root.val)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        if "(" not in data and ")" not in data:
            return TreeNode(int(data))
        begin = data.index("(")
        val = int(data[:begin])
        node = TreeNode(val)
        cnt = 1
        for j in range(begin+1,len(data)):
            if data[j]=="(":
                cnt+=1
            elif data[j]==")":
                cnt-=1
                if cnt==0:
                    break
        node.left = self.deserialize(data[begin+1:j])
        node.right = self.deserialize(data[j+2:-1])
        return node

File: 449 Serialize and Deserialize BST.go
// Serialization is converting a data structure or object into a sequence of bits so that it can be stored in a file or memory buffer, or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

// Design an algorithm to serialize and deserialize a binary search tree. There is no restriction on how your serialization/deserialization algorithm should work. You need to ensure that a binary search tree can be serialized to a string, and this string can be deserialized to the original tree structure.

// The encoded string should be as compact as possible.

 

// Example 1:

// Input: root = [2,1,3]
// Output: [2,1,3]
// Example 2:

// Input: root = []
// Output: []
 

// Constraints:

// The number of nodes in the tree is in the range [0, 104].
// 0 <= Node.val <= 104
// The input tree is guaranteed to be a binary search tree.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import "strconv"

type Codec struct {
    
}

func Constructor() Codec {
    return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
    if root == nil {
        return ""
    }
    return fmt.Sprintf("%d(%s)(%s)",root.Val,this.serialize(root.Left),this.serialize(root.Right))
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {    
    if len(data)==0 {
        return nil
    }
    cnt := 0
    left,right := 0, 0
    for i:=0;i<len(data);i++{
        if data[i]=='(' {
            cnt++
            if cnt==1 {
                left = i
            }
        }else if data[i]==')'{
            cnt--
            if cnt==0 {
                right = i
                break
            }
        }
    }
    curNode := &TreeNode{}
    val,_ := strconv.Atoi(data[:left])
    curNode.Val = val
    curNode.Left = this.deserialize(data[left+1:right])
    curNode.Right = this.deserialize(data[right+2:len(data)-1])
    return curNode
    
    
}


/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor()
 * deser := Constructor()
 * tree := ser.serialize(root)
 * ans := deser.deserialize(tree)
 * return ans
 */